smartsim/entity/strategies.py

```diff
@@ -117,108 +130,113 @@ def _make_sanitized_custom_strategy(
 
     @functools.wraps(fn)
     def _impl(
-        params: t.Mapping[str, t.Sequence[str]], n_permutations: int
-    ) -> list[dict[str, str]]:
+        params: t.Optional[t.Mapping[str, t.Sequence[str]]],
+        exe_args: t.Optional[t.Mapping[str, t.Sequence[t.Sequence[str]]]],
+        n_permutations: int = 0,
+    ) -> list[ParamSet]:
         try:
-            permutations = fn(params, n_permutations)
+            permutations = fn(params, exe_args, n_permutations)
         except Exception as e:
             raise errors.UserStrategyError(str(fn)) from e
         if not isinstance(permutations, list) or not all(
-            isinstance(permutation, dict) for permutation in permutations
+            isinstance(permutation, ParamSet) for permutation in permutations
         ):
             raise errors.UserStrategyError(str(fn))
         return permutations
 
     return _impl
 
 
-# create permutations of all parameters
-# single application if parameters only have one value
 @_register("all_perm")
 def create_all_permutations(
-    params: t.Mapping[str, t.Sequence[str]],
-    _n_permutations: int = 0,
-    # ^^^^^^^^^^^^^
-    # TODO: Really don't like that this attr is ignored, but going to leave it
-    #       as the original impl for now. Will change if requested!
-) -> list[dict[str, str]]:
+    file_params: t.Mapping[str, t.Sequence[str]],
+    exe_arg_params: t.Mapping[str, t.Sequence[t.Sequence[str]]],
+    n_permutations: int = 0,
+) -> list[ParamSet]:
     """Take a mapping parameters to possible values and return a sequence of
     all possible permutations of those parameters.
-
     For example calling:
-
     .. highlight:: python
     .. code-block:: python
-
         create_all_permutations({"A": ["1", "2"],
                                  "B": ["3", "4"]})
-
     Would result in the following permutations (not necessarily in this order):
-
     .. highlight:: python
     .. code-block:: python
-
         [{"A": "1", "B": "3"},
          {"A": "1", "B": "4"},
          {"A": "2", "B": "3"},
          {"A": "2", "B": "4"}]
-
     :param params: A mapping of parameter names to possible values
     :param _n_permutations: <ignored>
     :return: A sequence of mappings of all possible permutations
     """
-    permutations = itertools.product(*params.values())
-    return [dict(zip(params, permutation)) for permutation in permutations]
+    file_params_permutations = itertools.product(*file_params.values())
+    param_zip = (
+        dict(zip(file_params, permutation)) for permutation in file_params_permutations
+    )
+    exe_arg_params_permutations = itertools.product(*exe_arg_params.values())
+    exe_arg_zip = [
+        dict(zip(exe_arg_params, permutation))
+        for permutation in exe_arg_params_permutations
+    ]
```

Comment:
nit: consider making this a generator to keep this evaluation lazy
```suggestion
    exe_arg_zip = (
        dict(zip(exe_arg_params, permutation))
        for permutation in exe_arg_params_permutations
    )
```